### PR TITLE
Feat[BMQ, MQB]: shutdown v2, optimizing shutdown logic

### PR DESCRIPTION
**Pull request by dorjesinpo**
Changing shutdown logic to minimize control plane traffic.

1.  StopRequest is now per App (not Cluster)
2. Downstreams of the the shutting down do not deconfigure queues.  Upstreams do.
3. The shutting down waits for unconfirmed (not the downstreams)

```
DownStream              Shutting down broker:           Upstream

                            InitiateShutdown:
                
                            ClientSessions:
                                changes the state
                    
                            Broker sessions:
                    <--         StopRequest V2          -->

Clusters                                                    Clusters
and                                                             The ClusterNodeSession:
ClusterProxies (CQH):                                               QueueHandles:
    Queues:                                                             deconfigureAll
        buffer PUTs
                                                            ClusterProxies:
                                                                The ClientSession:
                                                                    QueueHandles:
                                                                        deconfigureAll
                    -->         StopResponse V2         <--
            
                            Clusters:
                                stopPushing
                                checkUnconfirmedV2
                        
                                continueShutdownDispatched
                                    change the state(s)
                                    cancelAllRequests
                                    delete (dirty) queues
```

### CODE REVIEW

**Review by pniedzielski — CHANGES_REQUESTED** [review]
Nothing big.  A few questions inline, a pervasive typo, and some comment/docs changes.

## Overall comments

  - Once we remove the v1 shutdown code, this will be a bit simpler.  I think then it might be worth taking a little time to look at simplifying the structure of some of the parts that we've touched here (the separation of labor between `initializeShutdown` vs `stop` in `mqba_application` won't really make sense at that stage).

  - Now that we're open-source, I think we need be a bit more careful about what versions of brokers users may try to run.  I noted that already we might want to turn off support for ungraceful shutdown, since all OSS brokers support it.  But I think we should be intentional about turning off graceful shutdown v1, document when we do it, and provide an upgrade process for this (in the manner of https://kafka.apache.org/documentation/#upgrade).

  - I found this comment in `mqbblp_clusterproxy`'s `ClusterProxy::stopDispatched`:
    ```
    // TBD:  In the future, we should review the entire shutdown logic, and
    //       eventually implement a request/response based mechanism in
    //       order to provide true graceful shutdown.
    ```
    A search revealed that this indeed was from before graceful shutdown logic v1, and from what I understand about this part of the code, this comment can safely be removed.  I think changing some of the `// temporary` comments to something greppable as I noted below will mitigate things like this being left behind from this PR.
    
## Merging

I think we'll need to unrevert Luke's admin API changes before we merge this in.  So either: you can make the changes I suggest here and I'll approve before we unrevert the changes, and I can approve again after you rebase, or you can hold off on making the changes until after we unvert those changes and I'll approve then.

File: src/groups/bmq/bmqp/bmqp_requestmanager.h
Comment (dorjesinpo) [reply src/groups/bmq/bmqp/bmqp_requestmanager.h:1095]: Yes, this is unrelated.  Some UT(s) were failing allocators check

File: src/groups/mqb/mqba/mqba_clientsession.h
Comment (dorjesinpo) [reply src/groups/mqb/mqba/mqba_clientsession.h:276]: Replacing with `// TODO(shutdown-v2): TEMPORARY, remove when all switch to StopRequest V2.`

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp:431]: Should we change it? How about `undoRouting`?

File: src/groups/mqb/mqba/mqba_application.cpp
Comment (dorjesinpo) [reply src/groups/mqb/mqba/mqba_application.cpp:580]: Agree, this method should return `false`, if there is a node not supporting `k_GRACEFUL_SHUTDOWN`.
This is more like a leftover from when shutdown v1 was implemented.  Looks like we need a cleanup, why don't we do V0 and v1 together ?

File: src/groups/mqb/mqba/mqba_application.cpp
Comment (dorjesinpo) [reply src/groups/mqb/mqba/mqba_application.cpp:494]: `DomainManager` does have a reference to `ClusterCatalog`.  Note this is `stop` method which does not destruct instances.
The  reason for the change is explained in the comment above.

File: src/groups/mqb/mqbi/mqbi_cluster.h
Comment (dorjesinpo) [reply src/groups/mqb/mqbi/mqbi_cluster.h:269]: It should be removed when everything switch to V2

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp
Comment (pniedzielski) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp]: `undoRouting` is a good name for this.  If you want to make that change in this PR or later, either way is okay by me.  Low priority.

File: src/groups/mqb/mqba/mqba_application.cpp
Comment (pniedzielski) [reply src/groups/mqb/mqba/mqba_application.cpp:556]: Doing V0 and V1 together sounds okay to me.

File: src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.h
Comment (pniedzielski) [reply src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.h]: Seems good to me.

File: src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp
Comment (pniedzielski) [reply src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp:5294]: Comment added below this line. ✅

File: src/groups/mqb/mqba/mqba_application.cpp
Comment (pniedzielski) [reply src/groups/mqb/mqba/mqba_application.cpp:470]: I see it now, yes.

**Review by pniedzielski — DISMISSED** [review]
👍🏻 

Let me know if you need a re-approval after fixing merge conflicts for getting this into `main`.

**Review by pniedzielski — APPROVED** [review]
Reapproving after rebase to fix merge conflicts.  Spot checked that no accidental changes were made.

**Review by pniedzielski — APPROVED** [review]

**Pull request merged**
